rust/agents/relayer/src/settings/matching_list.rs

```diff
@@ -189,39 +192,69 @@ struct WhitelistElement {
     destination_address: Filter<H256>,
 }
 
-impl Display for WhitelistElement {
+impl Display for ListElement {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
         write!(f, "{{sourceDomain: {}, sourceAddress: {}, destinationDomain: {}, destinationAddress: {}}}", self.source_domain, self.source_address, self.destination_domain, self.destination_address)
     }
 }
 
-impl Whitelist {
+#[derive(Copy, Clone, Debug)]
+struct MatchInfo<'a> {
+    src_domain: u32,
+    src_addr: &'a H256,
+    dst_domain: u32,
+    dst_addr: &'a H256,
+}
+
+impl<'a> From<&'a AbacusMessage> for MatchInfo<'a> {
+    fn from(msg: &'a AbacusMessage) -> Self {
+        Self {
+            src_domain: msg.origin,
+            src_addr: &msg.sender,
+            dst_domain: msg.destination,
+            dst_addr: &msg.recipient,
+        }
+    }
+}
+
+impl MatchingList {
+    pub fn msg_not_matches(&self, msg: &AbacusMessage) -> bool {
+        self.not_matches(msg.into())
+    }
+
     pub fn msg_matches(&self, msg: &AbacusMessage) -> bool {
-        self.matches(msg.origin, &msg.sender, msg.destination, &msg.recipient)
+        self.matches(msg.into())
     }
 
-    pub fn matches(
-        &self,
-        src_domain: u32,
-        src_addr: &H256,
-        dst_domain: u32,
-        dst_addr: &H256,
-    ) -> bool {
+    fn matches(&self, info: MatchInfo) -> bool {
         if let Some(rules) = &self.0 {
-            rules.iter().any(|rule| {
-                rule.source_domain.matches(&src_domain)
-                    && rule.source_address.matches(src_addr)
-                    && rule.destination_domain.matches(&dst_domain)
-                    && rule.destination_address.matches(dst_addr)
-            })
+            matches_any_rule(rules.iter(), info)
         } else {
             // by default if there is no whitelist, allow everything
             true
         }
     }
+
+    fn not_matches(&self, info: MatchInfo) -> bool {
+        if let Some(rules) = &self.0 {
+            matches_any_rule(rules.iter(), info)
+        } else {
+            // by default if there is no blacklist, disallow nothing
```

Comment:
I like the idea, but because of how this is getting brought in from serde I would rather come up with a different option for specifying the default rather than offloading it to the config file where whitelist or blacklist already canonically says which should be the default. So I think I might go for passing the default into the matches function OR just returning an option that can be used normally.

I agree that using the not_matches like that is a bit confusing.

Reply:
I agree with Trevor that this is easy to mis-use and a layering violation / pushing context-awareness into an implicit usage requirement (matches v non_matches)